src/mscorlib/shared/System/Diagnostics/Tracing/EventSource.cs

```diff
@@ -692,6 +689,35 @@ protected EventSource(EventSourceSettings settings, params string[] traits)
             Initialize(eventSourceGuid, eventSourceName, traits);
         }
 
+#if FEATURE_PERFTRACING
+        // Generate all EventPipeEvent instances which belong to this EventProvider
+        private unsafe void GenerateEventPipeEvent()
+        {
+            List<uint> allEvents = new List<uint>();
+            Debug.Assert(m_eventData != null);
+            int cnt = m_eventData.Length;
+            for (int i = 0; i < cnt; i++)
+            {
+                Int64 keywords = m_eventData[i].Descriptor.Keywords;
+                uint eventID = (uint)m_eventData[i].Descriptor.EventId;
+                uint eventVersion = m_eventData[i].Descriptor.Version;
+                uint level = m_eventData[i].Descriptor.Level;
+                // TODO: what's the value of needStack?
+                bool needStack = false;
+                
+                // Note: it's not clear that elements in m_eventData are unique or not
+                // so that we use a list allEvents to record the unique elements. 
+                if(allEvents.Contains(eventID))
```

Comment:
With the exception of ID=0 (which is an 'illegal' value), eventIDs need to be unique (EventSource does check for this because it is an easy mistake).   Thus you can rip out the allEvents (or leave it as debug-only code).    Some EventSources can have a large number of events (> 10K), so we really don't want something here that will make this N*N.